### PR TITLE
PERF-4147 - New workload to help report CPU cycle metrics

### PR DESCRIPTION
**Pull request by MAhmadShah**
**Jira Ticket:** [PERF-4147](https://jira.mongodb.org/browse/PERF-4147)

**Whats Changed:**  
We have added the CPYCycleMetrics workload that will help us report on cpu cycle metrics on the new Linux 3-Node ReplSet CPU Cycle Metrics 2023-06 variant.

**Patch testing results:**  
[Patch](https://spruce.mongodb.com/version/6480d13632f417f2ff943558/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC) showing the results of the new workload running on the new variant. Looking at the trend charts view, we can see the new metrics being reported.

**Related PRs:**   

- https://github.com/10gen/mongo/pull/13421
- https://github.com/10gen/dsi/pull/1309

### CODE REVIEW

**Review by SimonEismann — COMMENTED** [review]
Some feedback around the new tests

File: src/workloads/query/CPUCycleMetrics.yml [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  when utilizing the Linux 3-Node ReplSet CPU Cycle Metrics 2023-06 variant
+
+dbname: &dbname 10kDocs
+DocumentCount: &NumDocs 10000
+CollectionCount: &NumColls 1
+
+NumPhases: &NumPhases 3
+
+Document: &doc
+  a: {^RandomInt: {min: 0, max: *NumDocs}}
```
[the remark this answers is not in the document]
Comment (MAhmadShah) [reply src/workloads/query/CPUCycleMetrics.yml]: Good call, I've updated this to increment instead

File: src/workloads/query/CPUCycleMetrics.yml [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  b: {^RandomString: {length: 16}}
+
+Clients:
+  Default:
+    QueryOptions:
+      maxPoolSize: 1000
```
[the remark this answers is not in the document]
Comment (MAhmadShah) [reply src/workloads/query/CPUCycleMetrics.yml]: Not necessary, was put as a place holder and forgot to remove it. Gone now.

File: src/workloads/query/CPUCycleMetrics.yml [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    Type: CrudActor
+    Database: *dbname
+    Threads: {^Parameter: {Name: "Threads", Default: 1}}
+    Phases:
+      OnlyActiveInPhases:
+        Active: {^Parameter: {Name: "ActivePhases", Default: [2]}}
+        NopInPhasesUpTo: *NumPhases
+        PhaseConfig:
+          Repeat: 10000
+          RecordFailure: true
```
[the remark this answers is not in the document]
Comment (MAhmadShah) [reply src/workloads/query/CPUCycleMetrics.yml]: Yes good call, I've updated ThrowOnFailure to be true

File: src/workloads/query/CPUCycleMetrics.yml [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+Document: &doc
+  a: {^RandomInt: {min: 0, max: *NumDocs}}
+  b: {^RandomString: {length: 16}}
+
+Clients:
+  Default:
+    QueryOptions:
+      maxPoolSize: 1000
+
+ActorTemplates:
```
[the remark this answers is not in the document]
Comment (MAhmadShah) [reply src/workloads/query/CPUCycleMetrics.yml]: Done

File: src/workloads/query/CPUCycleMetricsMixed.yml [hunk 5]
```diff
@@ -0,0 +1,59 @@
+SchemaVersion: 2018-07-01
```
[the remark this answers is not in the document]
Comment (MAhmadShah) [reply src/workloads/query/CPUCycleMetricsMixed.yml:1]: My goal was to setup a workload where we can perform CRUD actions. So here the idea was to insert one document, update it to a new value and then delete it. We repeat this 10k times and watch the CPU cycle metrics.

**Review by SimonEismann — CHANGES_REQUESTED** [review]
As discussed via slack, let's split the second test into two to three separate tests.

File: src/workloads/query/CPUCycleMetrics.yml
Comment (SimonEismann) [inline src/workloads/query/CPUCycleMetrics.yml:1]: CPUCycleMetric.yml -> CPUCycleMetricFind.yml?

File: src/workloads/query/CPUCycleMetricsDelete.yml [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+          BatchSize: 100
+          Threads: 1
+          DocumentCount: *NumDocs
+          Database: *dbname
+          CollectionCount: *NumColls
+          Document: *doc
+          Indexes:
+          - keys: {a: 1}
+
+  - Name: UpdateActor
```
Comment (SimonEismann) [inline src/workloads/query/CPUCycleMetricsDelete.yml:38]: UpdateActor --> DeleteActor

File: src/workloads/query/CPUCycleMetricsUpdate.yml [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+          RecordFailure: true
+          ThrowOnFailure: true
+          CollectionCount: *NumColls
+          MetricsName: "load"
+          Operations:
+          - OperationName: updateOne
+            OperationMetricsName: updates
+            OperationCommand:
+                Filter: { a: { ^RandomInt: { min: 0, max: *NumDocs } } }
+                Update: { $set: { ^RandomInt: { min: 0, max: *NumDocs } } }
```
Comment (SimonEismann) [inline src/workloads/query/CPUCycleMetricsUpdate.yml:57]: We might want to update b instead of a.

Also, do we want to use inc to filter for an item?

**Review by SimonEismann — CHANGES_REQUESTED** [review]
Minor nits

**Comment by MAhmadShah** [comment]
@SimonEismann updated patch showing all new workloads can be found here: https://spruce.mongodb.com/version/648734461e2d17497dae0861/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC

**Review by SimonEismann — APPROVED** [review]
LGTM! Kinda cool to see that for the find workload, the secondaries have a much better CPI since they aren't doing any replication work.

**Pull request merged**